[Feat] Car 관련 API 개발

Young-Yeong opened the pull request:
- Car - 필요한 칼럼 추가 - Dynamic update 적용

- CarController - getAccidentListByCarNumber - Table 수정으로 삭제 - getRepairListByCarNumber - Table 수정으로 삭제 - postCar - POST /api/v1/branches/cars - request : CarRequestDto carRequestDto String adminUsername - response : Long - patchCar - PATCH /api/v1/branches/cars - request : CarRequestDto carRequestDto String adminUsername Long carId - response : Boolean - DELETE /api/v1/branches/cars - request : String adminUsername String carNumber

- CarEvent - 차량 수리, 사고내역 Entity

- CarEventController - getAccidentListByCarNumber - GET /api/v1/branches/cars/accidents - request : String carNumber
            - response : List<CarEventResponseDto>
    - getRepairListByCarNumber
        - GET /api/v1/branches/cars/repairs
            - request : String carNumber
            - response : List<CarEventResponseDto>

- CarEventDto - 차량 수리, 사고내역 DTO

- CarEventRepository - JPA 및 Custom레포지토리 상속

- CarEventRepositoryCustom - getRepairListByCarNumber - 차량번호로 수리내역 조회 - getAccidentListByCarNumber - 차량번호로 사고내역 조회 - deleteRepairListByCarId - 차량 ID로 수리내역 삭제 - deleteAccidentListByCarId - 차량 ID로 사고내역 삭제

- CarEventRepositoryImpl - getRepairListByCarNumber - 차량번호로 CarEvent 조회 후 수리내역인 레코드들 List<CarEventResponseDto>로 반환 - getAccidentListByCarNumber - 차량번호로 CarEvent 조회 후 사고내역인 레코드들 List<CarEventResponseDto>로 반환 - deleteRepairListByCarId - carId로 CarEvent 조회 후 수리내역인 레코드들 삭제 - deleteAccidentListByCarId - carId로 CarEvent 조회 후 사고내역인 레코드들 삭제

- CarEventResponseDto - 차량 사고, 수리내역 응답 DTO

- CarEventService - getRepairListByCarNumber - 차량번호로 수리내역 반환 - getAccidentListByCarNumber - 차량번호로 사고내역 반환

- CarRepositoryCustom - findAccidentListByCarNumber - Table 수정으로 이동 - findRepairListByCarNumber - Table 수정으로 이동

- CarRepositoryImpl - findAccidentListByCarNumber - Table 수정으로 이동 - findRepairListByCarNumber - Table 수정으로 이동 - searchUsableCar - where문 조건 수정

- CarRequestDto - 차량 등록, 변경하는 Dto

- CarService - getAccidentListByCarNumber - Table 수정으로 이동 - getRepairListByCarNumber - Table 수정으로 이동 - postCar - 요청받은 carRequestDto에 맞게 유효성 검사 후 차량 등록 - patchCar - 요청받은 carRequestDto에 맞게 유효성 검사 후 차량 변경 - deleteCar - 유효성 검사 후 carId에 해당하는 레코드 삭제

- ReservationDetailDto - drivers 추가

- ReservationListResponseDto - 필요한 요소들 추가

- ReservationRepositoryCustom - getAccidentListByUsername - 반환값 변경 - getRepairListByUsername - 반환값 변경 - findNowReservationDriversByUsername - 진행 중인 예약의 운전자 조회

- ReservationRepositoryImpl - getAccidentListByUsername - 반환값 변경 - getRepairListByUsername - 반환값 변경 - findNowReservationDriversByUsername - username으로 현재 예약건의 운전자 리스트 조회 - getReservationListByUsername - select 컬럼 변경 - findLatestReservationInfo - 조건 수정
- Driver - 운전자 관련 Entity

- UserController - getAccidentListByUsername - response값 수정 - getRepairListByUsername - response값 수정 

- Related to : #87

Young-Yeong committed:
Feat : Car 관련 API 개발

- Car
    - 필요한 칼럼 추가
    - Dynamic update 적용

- CarController
    - getAccidentListByCarNumber
        - Table 수정으로 삭제
    - getRepairListByCarNumber
        - Table 수정으로 삭제
    - postCar
        - POST /api/v1/branches/cars
            - request : CarRequestDto carRequestDto
                        String adminUsername
            - response : Long
    - patchCar
        - PATCH /api/v1/branches/cars
            - request : CarRequestDto carRequestDto
                        String adminUsername
                        Long carId
            - response : Boolean
        - DELETE /api/v1/branches/cars
            - request : String adminUsername
                        String carNumber

- CarEvent
    - 차량 수리, 사고내역 Entity

- CarEventController
    - getAccidentListByCarNumber
        - GET /api/v1/branches/cars/accidents
            - request : String carNumber
            - response : List<CarEventResponseDto>
    - getRepairListByCarNumber
        - GET /api/v1/branches/cars/repairs
            - request : String carNumber
            - response : List<CarEventResponseDto>

- CarEventDto
    - 차량 수리, 사고내역 DTO

- CarEventRepository
    - JPA 및 Custom레포지토리 상속

- CarEventRepositoryCustom
    - getRepairListByCarNumber
        - 차량번호로 수리내역 조회
    - getAccidentListByCarNumber
        - 차량번호로 사고내역 조회
    - deleteRepairListByCarId
        - 차량 ID로 수리내역 삭제
    - deleteAccidentListByCarId
        - 차량 ID로 사고내역 삭제

- CarEventRepositoryImpl
    - getRepairListByCarNumber
        - 차량번호로 CarEvent 조회 후 수리내역인 레코드들 List<CarEventResponseDto>로 반환
    - getAccidentListByCarNumber
        - 차량번호로 CarEvent 조회 후 사고내역인 레코드들 List<CarEventResponseDto>로 반환
    - deleteRepairListByCarId
        - carId로 CarEvent 조회 후 수리내역인 레코드들 삭제
    - deleteAccidentListByCarId
        - carId로 CarEvent 조회 후 사고내역인 레코드들 삭제

- CarEventResponseDto
    - 차량 사고, 수리내역 응답 DTO

- CarEventService
    - getRepairListByCarNumber
        - 차량번호로 수리내역 반환
    - getAccidentListByCarNumber
        - 차량번호로 사고내역 반환

- CarRepositoryCustom
    - findAccidentListByCarNumber
        - Table 수정으로 이동
    - findRepairListByCarNumber
        - Table 수정으로 이동

- CarRepositoryImpl
    - findAccidentListByCarNumber
        - Table 수정으로 이동
    - findRepairListByCarNumber
        - Table 수정으로 이동
    - searchUsableCar
        - where문 조건 수정

- CarRequestDto
    - 차량 등록, 변경하는 Dto

- CarService
    - getAccidentListByCarNumber
        - Table 수정으로 이동
    - getRepairListByCarNumber
        - Table 수정으로 이동
    - postCar
        - 요청받은 carRequestDto에 맞게 유효성 검사 후 차량 등록
    - patchCar
        - 요청받은 carRequestDto에 맞게 유효성 검사 후 차량 변경
    - deleteCar
        - 유효성 검사 후 carId에 해당하는 레코드 삭제

- ReservationDetailDto
    - drivers 추가

- ReservationListResponseDto
    - 필요한 요소들 추가

- ReservationRepositoryCustom
    - getAccidentListByUsername
        - 반환값 변경
    - getRepairListByUsername
        - 반환값 변경
    - findNowReservationDriversByUsername
        - 진행 중인 예약의 운전자 조회

- ReservationRepositoryImpl
    - getAccidentListByUsername
        - 반환값 변경
    - getRepairListByUsername
        - 반환값 변경
    - findNowReservationDriversByUsername
        - username으로 현재 예약건의 운전자 리스트 조회
    - getReservationListByUsername
        - select 컬럼 변경
    - findLatestReservationInfo
        - 조건 수정
- Driver
    - 운전자 관련 Entity

- UserController
    - getAccidentListByUsername
        - response값 수정
    - getRepairListByUsername
        - response값 수정
Related to : #87

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/controller/CarController.java:
```diff
@@ -2,15 +2,15 @@
 
 import java.util.List;
 
+@Validated
```
class에 붙이는 것 신기하네요

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/controller/CarEventController.java:
```diff
@@ -0,0 +1,33 @@
+        this.carEventService = carEventService;
+    }
+
+    @GetMapping("/accidents")
+    public ResponseEntity<List<CarEventResponseDto>> getAccidentListByCarNumber(@RequestParam String carNumber){
+        return ResponseEntity.ok(carEventService.getAccidentListByCarNumber(carNumber));
+    }
+
+    @GetMapping("/repairs")
+    public ResponseEntity<List<CarEventResponseDto>> getRepairListByCarNumber(@RequestParam String carNumber){
+        return ResponseEntity.ok(carEventService.getRepairListByCarNumber(carNumber));
+    }
```
`carController`와 분리한 것 좋네요 👍

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/dto/CarRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+@NoArgsConstructor
+@Builder
+public class CarRequestDto {
+    @NotNull(message = "차량번호가 비었습니다.")
+    private String carNumber;
+    @NotNull(message = "총 주행거리가 비었습니다.")
+    private Integer totalDistance;
+    @NotNull(message = "가격이 비었습니다.")
+    private Integer carPrice;
+    @NotNull(message = "할인율이 비었습니다.")
+    private Integer discountRate;
+    private String discountReason;
+    @NotNull(message = "차량 설명이 비었습니다.")
+    private String carDescription;
+    @NotNull(message = "차량이름이 비었습니다.")
+    private String carName;
+    @NotNull(message = "최대 탑승 인원이 비었습니다.")
+    private Integer maxPassenger;
+    @NotNull(message = "차량 크기가 비었습니다.")
+    private String carSize;
+    @NotNull(message = "유종이 비었습니다.")
+    private String oilType;
+    @NotNull(message = "구동기가 비었습니다.")
+    private String transmission;
+    private boolean isKorean;
+    @NotNull(message = "브랜드가 비었습니다.")
+    private String carBrand;
+    @NotNull(message = "차량 출시일이 비었습니다.")
```
message 작성해준 것 좋네요 조금 노가다 같긴한데...ㅋㅋㅋㅋㅋ

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/dto/CarRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+    private String carBrand;
+    @NotNull(message = "차량 출시일이 비었습니다.")
+    private LocalDateTime releaseDate;
+    private List<CarEventDto> repairList;
+    private List<CarEventDto> accidentList;
```
같은 이벤트인걸 보니 CarRepairEventDto, CarAccidentEventDto로 상속구조를 만들수도 있었을 것같아요
```java
public class CarEventDto {
    private String title;
    private LocalDateTime eventDate;
    private String content;
    private Boolean isRepair;
}
```

```java
@Getter
@ToString
@NoArgsConstructor
public class CarRepairEventDto extends CarEventDto {

    private String title;
    private LocalDateTime eventDate;
    private String content;

    @Builder.Default
    CarRepairEventDto(String title, LocalDateTime eventDate, String content) {
        super(title, eventDate, content, true);
    }
}
```

Capoomaru replied:
같은 이벤트인걸 보니 CarRepairEventDto, CarAccidentEventDto로 상속구조를 만들수도 있었을 것같아요
```java
public class CarEventDto {
    private String title;
    private LocalDateTime eventDate;
    private String content;
    private Boolean isRepair;
}
```

```java
@Getter
@ToString
@NoArgsConstructor
public class CarRepairEventDto extends CarEventDto {

    private String title;
    private LocalDateTime eventDate;
    private String content;

    @Builder.Default
    CarRepairEventDto(String title, LocalDateTime eventDate, String content) {
        super(title, eventDate, content, true);
    }
}
```

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/dto/CarRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+    private String carBrand;
+    @NotNull(message = "차량 출시일이 비었습니다.")
+    private LocalDateTime releaseDate;
+    private List<CarEventDto> repairList;
+    private List<CarEventDto> accidentList;
```
같은 이벤트인걸 보니 CarRepairEventDto, CarAccidentEventDto로 상속구조를 만들수도 있었을 것같아요
```java
public class CarEventDto {
    private String title;
    private LocalDateTime eventDate;
    private String content;
    private Boolean isRepair;
}
```

```java
@Getter
@ToString
@NoArgsConstructor
public class CarRepairEventDto extends CarEventDto {

    private String title;
    private LocalDateTime eventDate;
    private String content;

    @Builder
    CarRepairEventDto(String title, LocalDateTime eventDate, String content) {
        super(title, eventDate, content, true);
    }
}
```
```java
@Getter
@ToString
@NoArgsConstructor
public class CarAccidentEventDto extends CarEventDto {

    private String title;
    private LocalDateTime eventDate;
    private String content;

    @Builder
    CarAccidentEventDto(String title, LocalDateTime eventDate, String content) {
        super(title, eventDate, content, false);
    }
}
```

Capoomaru replied:
이제보니 이건 request이고, eventDto는 isRepair에 대한 항목이 없군요! 흠 어떻게 처리할지 애매하네요

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/entity/Car.java:
```diff
@@ -47,11 +49,41 @@ public class Car extends BaseTimeEntity {
+//    //에러 표기되지만 실제 동작에는 문제 없음
+//    @Column(name = "accident_list", columnDefinition = "text[]")
+//    private List<String> accidentList;
+    public Car updateCar(CarRequestDto carRequestDto, CarSpecification carSpecification, Branch branch) {
+        this.carSpec = carSpecification;
+        this.branch = branch;
+        this.carNumber = carRequestDto.getCarNumber();
+        this.totalDistance = carRequestDto.getTotalDistance();
+        this.carPrice = carRequestDto.getCarPrice();
+        this.discountRate = carRequestDto.getDiscountRate();
+        this.discountReason = carRequestDto.getDiscountReason();
+        this.carDescription = carRequestDto.getCarDescription();
+        return this;
+    }
```
@DynamicUpdate 를 안쓰시는 이유가 있을까요?

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/entity/CarEvent.java:
```diff
@@ -0,0 +1,45 @@
+    @Column(name = "car_event_id")
+    private Long carEventId;
+
+    @NotNull
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "car_id")
+    private Car car;
```
뭔가 주체가 반대인 것 같은데요?
car 내에 CarEventList가 있는게 아닐까요?

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/repository/CarEventRepositoryImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    public List<CarEventResponseDto> getRepairListByCarNumber(String carNumber) {
+        return queryFactory.select(Projections.constructor(CarEventResponseDto.class,
+                        carEvent.title,
+                        carEvent.eventDate,
+                        carEvent.content))
```
위에서 말한 CarRepairEventDto를 사용하면 아래와 같이 사용할 수 있겠네요
```java
return queryFactory.select(Projections.constructor(CarRepairEventDto.class,
                        carEvent.title,
                        carEvent.eventDate,
                        carEvent.content,
                        true ))
```

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/repository/CarEventRepositoryImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    public List<CarEventResponseDto> getAccidentListByCarNumber(String carNumber) {
+        return queryFactory.select(Projections.constructor(CarEventResponseDto.class,
+                        carEvent.title,
+                        carEvent.eventDate,
+                        carEvent.content))
```
위의 내용과 동일합니다

Capoomaru reviewed (commented):
리뷰완료!

Young-Yeong committed:
Fix : #95 수정

- Car
    - DynamicUpdate 적용

- CarAccidentEventDto
    - 차량 사고 내역 Dto
    - CarEventDto 상속

- CarRepairEventDto
    - 차량 수리 내역 Dto
    - CarEventDto 상속

- CarRequestDto
    - 반환값 수정
Related to : #95

Capoomaru reviewed (approved):
LGTM
